File: problems_list/0018.四数之和/0018-四数之和.cpp
typedef long long LL;
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        int n=nums.size();
        sort(nums.begin(), nums.end());
        vector<vector<int>> ans;
        for(int i=0; i<n; i++){
            if(i>0 && nums[i-1]==nums[i]) continue;
            for(int j=i+1; j<n; j++){
                if(j>i+1 && nums[j-1]==nums[j]) continue;
                LL tar=1LL*target-nums[i]-nums[j];
                for(int l=j+1, r=n-1; l<r; ){
                    int tmp=nums[l]+nums[r];
                    if(tmp==tar){
                        vector<int> tmpV={nums[i], nums[j], nums[l], nums[r]};
                        ans.push_back(tmpV);
                        l++, r--;
                        while(l<r && nums[l]==nums[l-1]) l++;
                        while(l<r && nums[r]==nums[r+1]) r--;
                    }else if(tmp<tar){
                        l++;
                        while(l<r && nums[l]==nums[l-1]) l++;
                    }else{
                        r--;
                        while(l<r && nums[r]==nums[r+1]) r--;
                    }
                }
            }
        }
        return ans;
    }
};

File: problems_list/0018.四数之和/0018-四数之和.js
var fourSum = function(nums, target) {
    const quadruplets = [];
    if (nums.length < 4) {
        return quadruplets;
    }
    nums.sort((x, y) => x - y);
    const length = nums.length;
    for (let i = 0; i < length - 3; i++) {
        if (i > 0 && nums[i] === nums[i - 1]) {
            continue;
        }
        if (nums[i] + nums[i + 1] + nums[i + 2] + nums[i + 3] > target) {
            break;
        }
        if (nums[i] + nums[length - 3] + nums[length - 2] + nums[length - 1] < target) {
            continue;
        }
        for (let j = i + 1; j < length - 2; j++) {
            if (j > i + 1 && nums[j] === nums[j - 1]) {
                continue;
            }
            if (nums[i] + nums[j] + nums[j + 1] + nums[j + 2] > target) {
                break;
            }
            if (nums[i] + nums[j] + nums[length - 2] + nums[length - 1] < target) {
                continue;
            }
            let left = j + 1, right = length - 1;
            while (left < right) {
                const sum = nums[i] + nums[j] + nums[left] + nums[right];
                if (sum === target) {
                    quadruplets.push([nums[i], nums[j], nums[left], nums[right]]);
                    while (left < right && nums[left] === nums[left + 1]) {
                        left++;
                    }
                    left++;
                    while (left < right && nums[right] === nums[right - 1]) {
                        right--;
                    }
                    right--;
                } else if (sum < target) {
                    left++;
                } else {
                    right--;
                }
            }
        }
    }
    return quadruplets;
};